streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                     }
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```

Comment:
You can get the time when `onPartitionsAssigned` is called and pass that through. It may not be perfect, but i think it will be close enough. You then just compare if `MAX_POLL_INTERVAL_MS` has passed.
Anyway, `rebalanceTimeMs` needs to be reset whenever a task is successfully created. Lets say that `MAX_POLL_INTERVAL_MS` is set to 400 and i have 5 tasks to create. 
Currently what would happen:
1. First task created
2. Sleep for 50
3. Second task created
4. Sleep for 100
5. Third task created
6. Sleep for 200
7. Fourth task created
8. Sleep for 400
9. break

The point being that we've wasted almost double `MAX_POLL_INTERVAL_MS` by sleeping when we didn't need to. So by the time we get to 8 - it is probably too late anyway.